GHC 8 compatibility for stable-2.16

apoikos opened the pull request:
This makes stable-2.16 build with GHC 8.0 - 8.4, without breaking pre-GHC 8 compatibility. Commits up to and including bb9ead6  are direct cherry-picks from master, the subsequent commits restore compatibility where it was broken.

Tested on Jessie (GHC 7.6), Stretch (GHC 8.0) and Buster (GHC 8.4) and it seems to work.

iustin commented on cabal/ganeti.template.cabal:
```diff
@@ -80,7 +80,7 @@ library
   if flag(htest)
     build-depends:
-        HUnit                         >= 1.2.4.2    && < 1.3
+        HUnit                         >= 1.2.4.2    && < 1.7
```
Change max HUnit version?

apoikos replied:
Yes, that's a typo and it was so in the original commit as well. I'll fix the description, thanks!

iustin commented:
Fixing CLA since they are all either cherry-picks from attic/master or authored by you.

apoikos commented on cabal/ganeti.template.cabal:
```diff
@@ -49,7 +49,7 @@ library
     , process                       >= 1.1.0.1
     , random                        >= 1.0.1.1
-    , template-haskell              >= 2.7.0.0
+    , template-haskell              >= 2.11.0.0
```
I know, actually I'm mentioning it at the bottom of the commit message. I can squash both commits in one if needed, but I chose to leave the first commit as-is because it's a cherry-pick (unfortunately as I used `git rebase -i --onto` the commits didn't get a `cherry-picked from` trailer).

iustin replied:
Initially I didn't get that from the commit message, the updated one all good. And in general, yes, don't squash as that loses history/context. Thanks!

apoikos commented:
@iustin: updated all commits with cherry-pick information and fixed the commit messages in a couple of spots. The CLA flag needs to be set again, I guess :)

apoikos commented:
Also updated NEWS and relaxed the hinotify upper bound to allow 0.4 as well.

iustin commented:
> @iustin: updated all commits with cherry-pick information and fixed the commit messages in a couple of spots. The CLA flag needs to be set again, I guess :)

Yay, cherry-pick information, thanks a lot! I made some mistakes earlier too, but I think explicitly marking the cherry-picks is very good habit. Thanks!

iustin commented on NEWS:
```diff
@@ -43,6 +43,17 @@ to these changes. This was done in a backwards-compatible way, however
+the build fail.
+
+This version introduces support for newer GHC versions and Ganeti is now
+known to build with GHC 8.0, 8.2 and 8.4. Additionally, compatibility
```
"known to build with GHC 8.0, 8.2 and 8.4" - this is at first read unclear if compat with 7.x too. I'd explicitly mention "and 8.4, on top of the current 7.8".

apoikos replied:
I rephrased the whole section to also cover the libraries, thanks! @iustin: would you mind setting the CLA flag so that we can merge this?

iustin replied:
Thanks, new version looks good, resolving.

iustin commented on test/hs/Test/Ganeti/TestCommon.hs:
```diff
@@ -118,15 +118,20 @@ import Numeric
 -- * Arbitrary orphan instances
 
+#if !MIN_VERSION_QuickCheck(2,8,0)
 instance (Ord k, Arbitrary k, Arbitrary a) => Arbitrary (M.Map k a) where
```
Adding comment here but it's a comment on this entire commit. This works as a stop-gap, but it's the wrong fix. Let's not delay this patch series because reviewing N commits is tiring, I'll send in the weekend a proper fix with explanation.

iustin commented:
CLA updated. Note that some of your recent commits are missing a signed-off-by (not sure if intentional or not, it would be good to keep it), and the TagSet arbitrary change which is suboptimal.

Please commit, as github interface for reviewing large patch sets is suboptimal ☺, and let's have further fixes separately.

apoikos commented:
Hm, merging and/or rebasing broke the CLA again :(

apoikos commented:
And since the CLA was gone anyway, I added the Signed-off-by trailers again. Won't touch anything else!

iustin commented:
Fixed CLA, again. Googlebot is not too smart…
